exercises/A2-A3/4-boxed-data/src/main.rs

```diff
@@ -0,0 +1,87 @@
+/// Below you find a small start of a data type modelling the abstract syntax tree for an expression,
+/// and a small evaluator function.
+///
+/// Please extend this evaluator in the following ways:
+///
+/// - Add support for multiplication and division
+///
+/// - We have added the the form "Summation(Vec<Expr>)", representing the sum of a list of expressions.
+/// Question: why can we get away with Vec<Expr> enough in that case, instead of Box<Vec<Expr>> ?
+///
+/// - EXTRA: Since division can fail, the function eval needs to return an Option<i64>, where None indicates that a division by
+///   zero has occurred. Can you change the code so that that errors are propagated correctly? (hint: use the ? syntax).
+
+#[derive(Debug)]
+enum Expr {
+    Const(i64),
+    Add(Box<Expr>, Box<Expr>),
+    Sub(Box<Expr>, Box<Expr>),
+    Var,
+    Summation(Vec<Expr>),
+}
+
+// inject these two identifiers directly into the current namespace
+use Expr::Const;
+use Expr::Summation;
+use Expr::Var;
+
+// These are convenience functions, so you don't have to type "Box::new" as often
+// when building test-data types
+fn add(x: Expr, y: Expr) -> Expr {
+    Expr::Add(Box::new(x), Box::new(y))
+}
+
+fn sub(x: Expr, y: Expr) -> Expr {
+    Expr::Sub(Box::new(x), Box::new(y))
+}
+
+fn mul(x: Expr, y: Expr) -> Expr {
+    todo!()
+}
+
+fn div(x: Expr, y: Expr) -> Expr {
+    todo!()
+}
+
+// ...
+
+fn eval(expr: &Expr, var: i64) -> i64 {
+    // this should return an Option<i64>
+    use Expr::*;
+    match expr {
+        Const(k) => *k,
+        Var => var,
+        Add(lhs, rhs) => eval(lhs, var) + eval(rhs, var),
+        Sub(lhs, rhs) => eval(lhs, var) - eval(rhs, var),
+
+        Summation(exprs) => {
+            let mut acc = 0;
+            for e in exprs {
+                acc += eval(e, var);
+            }
+            acc
+        }
+    }
+}
+
+fn main() {
+    let test = |expr| {
+        let value = rand::random::<i8>() as i64;
+        println!(
+            "{:?} with Var = {} ==> {}",
+            &expr,
+            value,
+            eval(&expr, value)
+        );
+    };
+
+    test(Const(5));
+    test(Var);
+    test(sub(Var, Const(5)));
+    test(add(sub(Var, Const(5)), Const(5)));
+    test(Summation(vec![Var, Const(1)]));
```

Comment:
make these actual tests?

Reply:
These are meant to show the thing in action, but adding actual tests is a good idea too!